tajo-plan/src/main/java/org/apache/tajo/plan/joinorder/GreedyHeuristicJoinOrderAlgorithm.java

```diff
@@ -57,17 +54,69 @@ public FoundJoinOrder findBestOrder(LogicalPlan plan, LogicalPlan.QueryBlock blo
     JoinEdge bestPair;
 
     while (remainRelations.size() > 1) {
+      Set<LogicalNode> checkingRelations = new LinkedHashSet<LogicalNode>();
+
+      for (LogicalNode relation : remainRelations) {
+        Collection <String> relationStrings = PlannerUtil.getRelationLineageWithinQueryBlock(plan, relation);
+        List<JoinEdge> joinEdges = new ArrayList<JoinEdge>();
+        String relationCollection = TUtil.collectionToString(relationStrings, ",");
+        List<JoinEdge> joinEdgesForGiven = joinGraph.getIncomingEdges(relationCollection);
+        if (joinEdgesForGiven != null) {
+          joinEdges.addAll(joinEdgesForGiven);
+        }
+        for (String relationString: relationStrings) {
+          joinEdgesForGiven = joinGraph.getIncomingEdges(relationString);
+          if (joinEdgesForGiven != null) {
+            joinEdges.addAll(joinEdgesForGiven);
+          }
+        }
+
+        // check if the relation is the last piece of outer join
+        boolean endInnerRelation = false;
+        for (JoinEdge joinEdge: joinEdges) {
```

Comment:
Actually, I thought that the order of JoinEdges in joinEdges list does not matter.
Since the given relation is the right relation for all the JoinEdges,
I just check the existence of JoinEdge that is outer join.
And, in my thought, the result does not vary regardless of the order of JoinEdge's occurrence.